cpp/ql/src/experimental/Security/CWE/CWE-754/ImproperCheckReturnValueScanf.ql

```diff
@@ -0,0 +1,108 @@
+    (
+      fc.getTarget().hasGlobalOrStdName("fscanf") or
+      fc.getTarget().hasGlobalOrStdName("sscanf") or
+      fc.getTarget().hasGlobalOrStdName("fscanf_s") or
+      fc.getTarget().hasGlobalOrStdName("sscanf_s")
+    ) and
+    result = 2
+  )
+}
+
+from FunctionCall fc, int n
+where
+  n = posArgumentInFunctionCall(fc) and
+  // Function return value is not evaluated.
+  fc instanceof ExprInVoidContext and
+  not isFromMacroDefinition(fc) and
+  exists(Expr e0, int i |
+    i in [n .. fc.getNumberOfArguments() - 1] and
+    // Fillable argument was not initialized.
+    (
+      fc.getArgument(i).(VariableAccess).getTarget() instanceof LocalScopeVariable or
+      fc.getArgument(i).(AddressOfExpr).getOperand().(VariableAccess).getTarget() instanceof
+        LocalScopeVariable
+    ) and
+    (
+      not fc.getArgument(i).(VariableAccess).getTarget().hasInitializer() and
+      not fc.getArgument(i)
+          .(AddressOfExpr)
+          .getOperand()
+          .(VariableAccess)
+          .getTarget()
+          .hasInitializer()
+    ) and
+    (
+      not fc.getArgument(i).(VariableAccess).getTarget().getAnAssignment().getASuccessor+() = fc and
+      not fc.getArgument(i)
+          .(AddressOfExpr)
+          .getOperand()
+          .(VariableAccess)
+          .getTarget()
+          .getAnAssignment()
+          .getASuccessor+() = fc
+    ) and
+    // After the call, the completed arguments are assigned or returned as the result of the operation of the upper function.
+    fc.getASuccessor+() = e0 and
+    (
+      (
+        e0.(Assignment).getRValue().(VariableAccess).getTarget() =
+          fc.getArgument(i).(AddressOfExpr).getOperand().(VariableAccess).getTarget() or
+        e0.(Assignment).getRValue().(VariableAccess).getTarget() =
+          fc.getArgument(i).(VariableAccess).getTarget()
+      )
+      or
+      e0.getEnclosingStmt() instanceof ReturnStmt and
+      (
+        e0.(VariableAccess).getTarget() =
+          fc.getArgument(i).(AddressOfExpr).getOperand().(VariableAccess).getTarget() or
+        e0.(VariableAccess).getTarget() = fc.getArgument(i).(VariableAccess).getTarget()
+      )
+      or
+      not exists(Expr e1 |
+        fc.getASuccessor+() = e1 and
+        (
+          e1.(VariableAccess).getTarget() =
+            fc.getArgument(i).(AddressOfExpr).getOperand().(VariableAccess).getTarget() or
+          e1.(VariableAccess).getTarget() = fc.getArgument(i).(VariableAccess).getTarget()
+        )
+      )
+    )
+  ) and
+  // After the call, filled arguments are not evaluated.
+  not exists(Expr e0, int i |
+    i in [n .. fc.getNumberOfArguments() - 1] and
+    fc.getASuccessor+() = e0 and
+    e0.getEnclosingElement() instanceof ComparisonOperation and
+    (
+      e0.(VariableAccess).getTarget() = fc.getArgument(i).(VariableAccess).getTarget() or
+      e0.(VariableAccess).getTarget() =
+        fc.getArgument(i).(AddressOfExpr).getOperand().(VariableAccess).getTarget()
+    )
+  )
```

Comment:
Could you clarify what this achieves? Like everything starting from line 71, this seems to also look at how the argument is being used, so I'm not sure why this separate from what I see there.